diff --git a/kernel/core/init.c b/kernel/core/init.c
index db10afa7..94db9eaa 100644
--- a/kernel/core/init.c
+++ b/kernel/core/init.c
@@ -24,6 +24,9 @@
 #include "feature/adb_root.h"
 #include "feature/selinux_hide.h"
 #include "feature/sulog.h"
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs.h>
+#endif
 #include "infra/symbol_resolver.h"
 
 #if defined(__x86_64__) && !defined(CONFIG_KSU_X86_PATCH_SYSCALL_DISPATCHER)
@@ -142,6 +145,11 @@ int __init kernelsu_init(void)
 
 	ksu_supercalls_init();
 
+#ifdef CONFIG_KSU_SUSFS
+	susfs_init();
+	pr_info("susfs enabled.\n");
+#endif
+
 	if (ksu_late_loaded) {
 		pr_info("late load mode, skipping kprobe hooks\n");
 
diff --git a/kernel/feature/kernel_umount.c b/kernel/feature/kernel_umount.c
index dd741202..32533897 100644
--- a/kernel/feature/kernel_umount.c
+++ b/kernel/feature/kernel_umount.c
@@ -1,3 +1,6 @@
 #include <linux/sched.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs_def.h>
+#endif
 #include <linux/slab.h>

@@ -83,6 +83,11 @@ int ksu_handle_umount(uid_t old_uid, uid_t new_uid)
 		return 0;
 	}
 
+#ifdef CONFIG_KSU_SUSFS
+	if (susfs_is_current_proc_umounted())
+		return 0;
+#endif
+
 	// There are 6 scenarios:
 	// 1. Normal app: zygote -> appuid
 	// 2. Isolated process forked from zygote: zygote -> isolated_process
diff --git a/kernel/feature/sucompat.c b/kernel/feature/sucompat.c
index 553cb6ae..288cd5f5 100644
--- a/kernel/feature/sucompat.c
+++ b/kernel/feature/sucompat.c
@@ -8,7 +8,10 @@
 #include <linux/uaccess.h>
 #include <asm/current.h>
 #include <linux/cred.h>
-#include <linux/fs.h>
+#include <linux/susfs_def.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs.h>
+#endif
 #include <linux/types.h>
 #include <linux/version.h>
 #include <linux/sched/task_stack.h>
@@ -103,6 +106,10 @@ long ksu_handle_faccessat_sucompat(int orig_nr, struct pt_regs *regs)
 	strncpy_from_user_nofault(path, *filename_user, sizeof(path));
 
 	if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
+#ifdef CONFIG_KSU_SUSFS
+		if (susfs_is_current_proc_umounted())
+			goto do_orig_facessat;
+#endif
 		old_cred = override_creds(ksu_cred);
 		if (is_ksud_exists()) {
 			ksu_compat_sulog('a');
@@ -139,6 +146,10 @@ long ksu_handle_stat_sucompat(int orig_nr, struct pt_regs *regs)
 	strncpy_from_user_nofault(path, *filename_user, sizeof(path));
 
 	if (unlikely(!memcmp(path, su_path, sizeof(su_path)))) {
+#ifdef CONFIG_KSU_SUSFS
+		if (susfs_is_current_proc_umounted())
+			goto do_orig_stat;
+#endif
 		old_cred = override_creds(ksu_cred);
 		if (is_ksud_exists()) {
 			ksu_compat_sulog('s');
@@ -186,8 +197,19 @@ long ksu_handle_execve_sucompat(const char __user **filename_user, int orig_nr,
 		goto do_orig_execve;
 	}
 
-	if (likely(memcmp(path, su_path, sizeof(su_path))))
+	if (likely(memcmp(path, su_path, sizeof(su_path)))) {
+#ifdef CONFIG_KSU_SUSFS
+		if (likely(!strstr(path, "/app_process") && !strstr(path, "/adbd"))) {
+			susfs_set_current_proc_umounted();
+		}
+#endif
+		goto do_orig_execve;
+	}
+
+#ifdef CONFIG_KSU_SUSFS
+	if (susfs_is_current_proc_umounted())
 		goto do_orig_execve;
+#endif
 
 	ksu_compat_sulog('x');
 	pr_info("sys_execve su found\n");
diff --git a/kernel/hook/setuid_hook.c b/kernel/hook/setuid_hook.c
index 3cccc780..839d4545 100644
--- a/kernel/hook/setuid_hook.c
+++ b/kernel/hook/setuid_hook.c
@@ -1,3 +1,6 @@
 #include <linux/compiler.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs_def.h>
+#endif
 #include <linux/version.h>

@@ -53,6 +53,13 @@ int ksu_handle_setresuid(uid_t old_uid, uid_t new_uid)
     // Handle kernel umount
     ksu_handle_umount(old_uid, new_uid);
 
+#ifdef CONFIG_KSU_SUSFS
+    extern struct work_struct susfs_extra_works;
+    if (!work_pending(&susfs_extra_works))
+        schedule_work(&susfs_extra_works);
+    susfs_set_current_proc_umounted();
+#endif
+
     return 0;
 }
 
diff --git a/kernel/hook/syscall_event_bridge.c b/kernel/hook/syscall_event_bridge.c
index f866a498..dfabef57 100644
--- a/kernel/hook/syscall_event_bridge.c
+++ b/kernel/hook/syscall_event_bridge.c
@@ -18,6 +18,9 @@
 #include "hook/syscall_hook.h"
 #include "hook/syscall_event_bridge.h"
 #include "feature/adb_root.h"
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs_def.h>
+#endif
 
 static int ksu_handle_init_mark_tracker(const char __user **filename_user)
 {
@@ -42,6 +45,9 @@ static int ksu_handle_init_mark_tracker(const char __user **filename_user)
     } else if (likely(strstr(path, "/app_process") == NULL && strstr(path, "/adbd") == NULL)) {
         pr_info("hook_manager: unmark %d exec %s\n", current->pid, path);
         ksu_clear_task_tracepoint_flag_if_needed(current);
+#ifdef CONFIG_KSU_SUSFS
+        susfs_set_current_proc_umounted();
+#endif
     }
 
     return 0;
diff --git a/kernel/supercall/dispatch.c b/kernel/supercall/dispatch.c
index 902631d9..da3d0801 100644
--- a/kernel/supercall/dispatch.c
+++ b/kernel/supercall/dispatch.c
@@ -5,6 +5,9 @@
 #include <linux/uaccess.h>
 #include <linux/version.h>
 #include <linux/thread_info.h>
+#ifdef CONFIG_KSU_SUSFS
+#include <linux/susfs.h>
+#endif
 #include "uapi/supercall.h"
 #include "supercall/internal.h"
 #include "arch.h" // IWYU pragma: keep
@@ -128,6 +131,9 @@ static int do_report_event(void __user *arg)
             } else {
                 pr_info("boot_complete triggered\n");
                 on_boot_completed();
+#ifdef CONFIG_KSU_SUSFS
+                susfs_start_sdcard_monitor_fn();
+#endif
             }
         }
         break;
@@ -954,6 +960,74 @@ long ksu_supercall_handle_ioctl(unsigned int cmd, void __user *argp)
     pr_info("ksu ioctl: cmd=0x%x from uid=%d\n", cmd, current_uid().val);
 #endif
 
+#ifdef CONFIG_KSU_SUSFS
+    switch (cmd) {
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+    case CMD_SUSFS_ADD_SUS_PATH:
+        susfs_add_sus_path(&argp);
+        return 0;
+    case CMD_SUSFS_ADD_SUS_PATH_LOOP:
+        susfs_add_sus_path_loop(&argp);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+    case CMD_SUSFS_HIDE_SUS_MNTS_FOR_NON_SU_PROCS:
+        susfs_set_hide_sus_mnts_for_non_su_procs(&argp);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_KSTAT
+    case CMD_SUSFS_ADD_SUS_KSTAT:
+        susfs_add_sus_kstat(&argp);
+        return 0;
+    case CMD_SUSFS_UPDATE_SUS_KSTAT:
+        susfs_update_sus_kstat(&argp);
+        return 0;
+    case CMD_SUSFS_ADD_SUS_KSTAT_STATICALLY:
+        susfs_add_sus_kstat(&argp);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
+    case CMD_SUSFS_SET_UNAME:
+        susfs_set_uname(&argp);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_ENABLE_LOG
+    case CMD_SUSFS_ENABLE_LOG:
+        susfs_enable_log(&argp);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+    case CMD_SUSFS_SET_CMDLINE_OR_BOOTCONFIG:
+        susfs_set_cmdline_or_bootconfig(&argp);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_OPEN_REDIRECT
+    case CMD_SUSFS_ADD_OPEN_REDIRECT:
+        susfs_add_open_redirect(&argp);
+        return 0;
+#endif
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
+    case CMD_SUSFS_ADD_SUS_MAP:
+        susfs_add_sus_map(&argp);
+        return 0;
+#endif
+    case CMD_SUSFS_ENABLE_AVC_LOG_SPOOFING:
+        susfs_set_avc_log_spoofing(&argp);
+        return 0;
+    case CMD_SUSFS_SHOW_ENABLED_FEATURES:
+        susfs_get_enabled_features(&argp);
+        return 0;
+    case CMD_SUSFS_SHOW_VARIANT:
+        susfs_show_variant(&argp);
+        return 0;
+    case CMD_SUSFS_SHOW_VERSION:
+        susfs_show_version(&argp);
+        return 0;
+    default:
+        break;
+    }
+#endif
+
     for (i = 0; ksu_ioctl_handlers[i].handler; i++) {
         if (cmd == ksu_ioctl_handlers[i].cmd) {
             // Check permission first
